### ISSUE TITLE
Parallel routes are rendered unnecessarily

### ISSUE THREAD

**Issue by slvstr-dev** [vercel/next.js#53292]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.16.0
      npm: 9.8.1
      Yarn: 1.22.19
      pnpm: 8.6.10
    Relevant Packages:
      next: 13.4.13-canary.5
      eslint-config-next: 13.4.12
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/slvstr-dev/test-parallel

### To Reproduce

- Install project
- Run project locally
- Check console for logs of both routes

### Describe the Bug

In `layout.tsx` I want to render conditionally render a parallel route, f.e. if a user is logged in or not. Currently it runs code from both routes. Console.log's of both `page.tsx` in root and `page.tsx` in `@auth` are being run.

### Expected Behavior

Based on the condition I only expect the console.log from the returned page to run.

**Comment by veiico** [comment]
Same issue here, just wanted to create an issue my self, but will leave this simple stackblitz example instead 😅 
https://stackblitz.com/edit/vercel-next-js-y3y9qb?file=app%2Fpage.tsx

**Comment by octet-stream** [comment]
Same issue, but I have different structure: within `/admin` route there are `@login` and `@admin` folders. On the `/admin/@admin/page.tsx` page I fetch some user-specific data to display that data on the page. The logic is the same: if user is logged in, then render `@admin`, otherwise `@login`. Then if I open `/admin` in browser, the data for `@admin` is being fetched no matter if I'm logged in or not.

As a workaround I moved data fetching to `admin/layout.tsx` and made `admin/@admin/page.tsx` a client component to be able to access the data via React Context. It works, but it's not ideal.

**Comment by chrisgrabinski** [comment]
I encountered the same problem in my app, where I used parallel/conditional routes to render the correct template for a given route. While trying to come up with a solution, I started to feel like there was no real reason for me to even use parallel routes in the first place. 

I ended up choosing a similar approach to @octet-stream, but without using client components and context. Instead, I simply do some minor fetching in `page.tsx` and then conditionally render a template component (RSC) which receives a `uri` prop. This is then used to fetch additional data directly in the template itself.

This solution may not work for everyone, but if you're only using parallel routes because it's [the "official" solution for conditional routes from the Next.js docs](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#conditional-routes), then it might be worth considering.

If you need any additional features from parallel routes, then another workaround I found is to add some logic to the `page.tsx` for your slot that returns `null` if it shouldn't be used. You should be able to reuse the same logic from your `layout.tsx` and when using `fetch` from Next.js, this will not lead to additional requests. 👌

I forked @veiico's great reproduction to showcase what I mean.
https://stackblitz.com/edit/vercel-next-js-mdb443?file=app%2Fpage.tsx

**Comment by tinleym** [comment]
@leerob sorry for the ping, but it would be useful to know if this is a bug or an expected behavior. In some cases this can be harmless, but if a conditional route is used for auth (per the example in the docs: https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#conditional-routes), this can lead to a lot of extra execution (and is probably not a good choice).

**Comment by kyeo76** [comment]
experiencing the same issue. i'm using next v13.5.4

**Comment by elzako** [comment]
Still an issue in v14.0.0

**Comment by springyboy** [comment]
Still and issue in 14.0.1.

**Comment by ShakurOo** [comment]
Any update ? This powerful feature should be fixed ...

**Comment by samcx** [comment]
@slvstr-dev Thank you for submitting an issue!

I can confirm this is still an issue with the implementation of Parallel routes. We will let you know when this is <picture data-single-emoji=":fixed:" title=":fixed:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/fixed/66bb8455f24bd50e.png" alt=":fixed:" align="absmiddle"></picture>!

**Comment by Nhollas** [comment]
6+ months, damn.

**Comment by payamtrack** [comment]
Having a similar issue, is there any update or any plan to fix it?

**Comment by samcx** [comment]
@payamtrack There is a plan to address this! Unfortunately, it is not an easy fix, so the team has been discussing this internally.

**Comment by payamtrack** [comment]
@samcx That's great, thanks for the update!

**Comment by nicnocquee** [comment]
@samcx 

I'm not sure if I should open another issue, but I encountered similar problem, a.k.a, the excessive rendering of server components in parallel routes, **even without conditionally rendering the slots**.

How to reproduce the bug:

- Clone demo repo: https://github.com/nicnocquee/nextjs-parallel-route-bug
- Run dev
- Open browser and go to http://localhost:3000/de
- Refresh the page
- Click the About link
- See in the terminal that the server components are all rendered multiple times.

What makes thing worst is that React's `cache` doesn't work in this case. The function being cached will be called as many times as the components are rendered (see that the `console.log` output inside the [_getData function](https://github.com/nicnocquee/nextjs-parallel-route-bug/blob/011a4e72e6e845502e98868e58187ff89fde19c0/app/%5Blng%5D/data.ts#L4) is called multiple times instead of just once. If the function queries the database, this would use resources unnecessarily.

**Comment by samcx** [comment]
@nicnocquee Please open a separate issue if it's not the exact issue described in this issue.

**Comment by ppiwo** [comment]
Any chance the docs could at least be updated to warn users that this feature is broken?
https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#conditional-routes

**Comment by samcx** [comment]
@ppiwo Of course! Maybe you can look to create a <picture data-single-emoji=":pr:" title=":pr:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/pr/876af7e7d65e1595.png" alt=":pr:" align="absmiddle"></picture> for this? → https://nextjs.org/docs/community/contribution-guide

**Comment by nuclei272** [comment]
I came across the same problem. I do understand that what's being returned to the client is still the correct layout and page, but it's a weird behavior. Not sure if this also happens in production. Is there any light and update on this issue?

**Comment by pulgueta** [comment]
Any updates as of now?

**Comment by ShakurOo** [comment]
Same in V15.0.1 ...

**Comment by ilbertt** [comment]
Are there any updates on this? I'm on v14.2.5 and I'm still facing the issue

**Comment by nachodeh** [comment]
We have a layout with a grid of items and a parallel route that displays product details (after clicking on the item). This issue is causing tons of unnecessary function invocations on prod, as all parallel routes seem to be called every time the main grid is loaded. 

Any updates here?

**Comment by ShakurOo** [comment]
Two years and half later, no news, look's not good ... @samcx

**Comment by CentralHaat** [comment]
@vercel This seems like a huge issue. Any update or timeline?

**Comment by justAnArthur** [comment]
Based on current the role-based auth cannot be done via parallel conditional routes.

**Comment by justAnArthur** [comment]
@samcx  Currently, role-based authentication cannot be implemented using parallel conditional routes in Next.js. I understand this limitation arises because pages and layouts are rendered in parallel, making it "imposible" to conditionally render routes based on roles (from layout).

However, this issue has existed for several years and is still documented in a way that may lead to misconceptions. This is problematic and cannot be overlooked.

I would like to propose a potential solution: introducing an "experimental" or "unstable" Next.js-standart directive for layouts - an exported function that evaluates which parallel route(s) to render (or both). From my perspective, this would be a simple and effective approach.

_Although I have not contributed to Next.js before, I am willing to open a PR implementing this idea if it aligns with the project’s direction, or to learn if there are deeper technical reasons why this solution would not work._

Unfortunately, due to this limitation, I currently have to migrate my code to an alternative solution for role-based authentication.

**Comment by jonathangraf** [comment]
I wrote a wrapper for slots, which utilizes useSelectedLayoutSegments to at least mitigate the rendering issue on the client. Server side code is still running so it doesn't solve anything. But if it's a low traffic route it shouldn't affect you to bad. Just make sure you don't fetch anything server side or run anything to complex than it's more or less fine.
